File: Java/src/org/hpms/mw/distcrud/ParticipantImpl.java
package org.hpms.mw.distcrud;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.NetworkInterface;
import java.net.ProtocolFamily;
import java.net.StandardProtocolFamily;
import java.net.StandardSocketOptions;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.function.Supplier;

final class ParticipantImpl implements IParticipant, IProtocol {

   private final ByteBuffer                        _header          = ByteBuffer.allocate( HEADER_SIZE );
   private final ByteBuffer                        _payload         = ByteBuffer.allocate( PAYLOAD_SIZE );
   private final ByteBuffer                        _message         = ByteBuffer.allocate( 64*1024 );
   private final Cache[]                           _caches          = new Cache[256];
   private final Dispatcher                        _dispatcher      = new Dispatcher( this );
   private final Map<ClassID, Supplier<Shareable>> _localFactories  = new TreeMap<>();
   private final Map<ClassID, ICRUD >              _remoteFactories = new TreeMap<>();
   private final Map<Integer, ICallback>           _callbacks       = new HashMap<>();
   private final int                               _publisherId;
   private final InetSocketAddress                 _target;
   private final DatagramChannel                   _out;
   private /* */ int                               _callId     = 1;
   private /* */ byte                              _cacheCount = 0;

   ParticipantImpl( int publisherId, InetSocketAddress group, NetworkInterface  intrfc ) throws IOException {
      final ProtocolFamily family =
         ( group.getAddress().getAddress().length > 4 )
            ? StandardProtocolFamily.INET6
            : StandardProtocolFamily.INET;
      _publisherId = publisherId;
      _target      = group;
      _out         = DatagramChannel
         .open     ( family )
//         .setOption( StandardSocketOptions.SO_REUSEADDR, true )
//         .bind     ( group )
         .setOption( StandardSocketOptions.IP_MULTICAST_IF, intrfc )
      ;
      createCache();
      System.out.printf( "Sending to %s via interface %s\n", group, intrfc );
   }

   @Override
   public void listen( NetworkInterface via, InetSocketAddress...others ) throws IOException {
      for( final InetSocketAddress other : others ) {
         new NetworkReceiver( this, other, via );
      }
   }

   int getPublisherId() {
      return _publisherId;
   }

   Shareable newInstance( ClassID classId, ByteBuffer frame ) {
      synchronized( _localFactories ) {
         final Supplier<Shareable> factory = _localFactories.get( classId );
         if( factory == null ) {
            return null;
         }
         final Shareable item = factory.get();
         item.unserialize( frame );
         return item;
      }
   }

   @Override
   public void registerLocalFactory( ClassID id, Supplier<Shareable> factory ) {
      synchronized( _localFactories ) {
         _localFactories.put( id, factory );
      }
   }

   @Override
   public void registerRemoteFactory( ClassID id, ICRUD publisher ) {
      synchronized( _remoteFactories ) {
         _remoteFactories.put( id, publisher );
      }
   }

   @Override
   public ICache getDefaultCache() {
      synchronized( _caches ) {
         return _caches[0];
      }
   }

   @Override
   public ICache createCache() {
      synchronized( _caches ) {
         return _caches[_cacheCount++] = new Cache( this );
      }
   }

   @Override
   public ICache getCache( byte cacheIndex ) {
      synchronized( _caches ) {
         return _caches[cacheIndex];
      }
   }

   @Override
   public IDispatcher getDispatcher() {
      return _dispatcher;
   }

   private void pushCreateOrUpdateItem( Shareable item ) throws IOException {
      synchronized( _out ){
         _payload.clear();
         item.serialize( _payload );
         _payload.flip();
         final int size = _payload.remaining();
         _header.clear();
         _header.put( SIGNATURE );
         _header.put((byte)FrameType.DATA_CREATE_OR_UPDATE.ordinal());
         _header.putInt( size );
         item._id   .serialize( _header );
         item._class.serialize( _header );
         _header.flip();
         _message.clear();
         _message.put( _header  );
         _message.put( _payload );
         _message.flip();
         _out.send( _message, _target );
      }
   }

   void publishUpdated( Set<? extends Shareable> updated ) throws IOException {
      for( final Shareable item : updated ) {
         pushCreateOrUpdateItem( item );
      }
   }

   private void pushDeleteItem( Shareable item ) throws IOException {
      synchronized( _out ){
         _header.clear();
         _header.put( SIGNATURE );
         _header.put((byte)FrameType.DATA_DELETE.ordinal());
         item._id.serialize( _header );
         _header.flip();
         _out.send( _header, _target );
      }
   }

   void publishDeleted( Set<? extends Shareable> deleted ) throws IOException {
      for( final Shareable item : deleted ) {
         pushDeleteItem( item );
      }
   }

   void call( String intrfcName, String opName, Arguments args, int callId ) throws IOException {
      synchronized( _out ) {
         final byte count = ( args == null ) ? 0 : (byte)args.getCount();
         _message.clear();
         _message        .put      ( SIGNATURE );
         _message        .put      ((byte)FrameType.OPERATION.ordinal());
         SerializerHelper.putString( intrfcName, _message );
         SerializerHelper.putString( opName    , _message );
         _message        .putInt   ( callId );
         _message        .put      ( count );
         if( args != null ) {
            args.serialize( _message );
         }
         _message.flip();
         _out.send( _message, _target );
      }
   }

   void call( String intrfcName, String opName, Arguments args, ICallback callback ) throws IOException {
      assert callback != null;
      call( intrfcName, opName, args, _callId );
      _callbacks.put( _callId++, callback );
   }

   void dataDelete( GUID id ) {
      synchronized( _caches ) {
         for( int i = 0; i < _cacheCount; ++i ) {
            _caches[i].deleteFromNetwork( id );
         }
      }
   }

   void dataUpdate( ByteBuffer frame, int payloadSize ) {
      synchronized( _caches ) {
         for( int i = 0; i < _cacheCount; ++i ) {
            final byte[] copy = new byte[payloadSize];
            System.arraycopy( frame.array(), frame.position(), copy, 0, payloadSize );
            final ByteBuffer fragment = ByteBuffer.wrap( copy );
//            Dump.dump( fragment );
            _caches[i].updateFromNetwork( fragment );
         }
      }
   }

   void callback( String intrfcName, String opName, Arguments args, int callId ) {
      final ICallback callback = _callbacks.get( callId );
      if( callback == null ) {
         System.err.printf( "Unknown Callback received: %s.%s, id: %d\n",
            intrfcName, opName, -callId );
      }
      else {
         callback.callback( intrfcName, opName, args );
      }
   }

   public boolean create( ClassID clsId, Arguments how ) throws IOException {
      if( clsId == null ) {
         return false;
      }
      final ICRUD publisher = _remoteFactories.get( clsId );
      if( publisher == null ) {
         return false;
      }
      publisher.create( how );
      return true;
   }

   public boolean update( GUID id, Arguments how ) throws IOException {
      for( int i = 0; i < _cacheCount; ++i ) {
         final Shareable item = _caches[i].read( id );
         if( item != null ) {
            final ICRUD publisher = _remoteFactories.get( item._class );
            if( publisher == null ) {
               return false;
            }
            publisher.update( item, how );
            return true;
         }
      }
      return false;
   }

   public boolean delete( GUID id ) throws IOException {
      for( int i = 0; i < _cacheCount; ++i ) {
         final Shareable item = _caches[i].read( id );
         if( item != null ) {
            final ICRUD publisher = _remoteFactories.get( item._class );
            if( publisher == null ) {
               return false;
            }
            publisher.delete( item );
            return true;
         }
      }
      return false;
   }
}

File: cpp/src/dcrud/ParticipantImpl.cpp
#include "ParticipantImpl.hpp"
#include "Dispatcher.hpp"
#include "Cache.hpp"
#include "IProtocol.hpp"
#include "NetworkReceiver.hpp"

#include <io/NetworkInterfaces.h>

#include <util/check.hpp>

#include <dcrud/Shareable.hpp>
#include <dcrud/ICallback.hpp>
#include <dcrud/ICRUD.hpp>
#include <dcrud/IRegistry.hpp>

#include <stdio.h>
#include <string.h>

#include <stdexcept>
#include <algorithm>

using namespace dcrud;

const byte ParticipantImpl::SIGNATURE[] = {'D','C','R','U','D'};

ParticipantImpl::ParticipantImpl(
   unsigned int                  publisherId,
   const io::InetSocketAddress & addr,
   const std::string &           intrfc )
 :
   _publisherId( publisherId  ),
   _header     ( HEADER_SIZE  ),
   _payload    ( PAYLOAD_SIZE ),
   _message    ( 64*1024 ), // UDP MAX packet size
   _outMutex   (),
   _out        ( 0 ),
   _cachesMutex(),
   _factoriesMutex(),
   _localFactories(),
   _remoteFactories(),
   _callbacks(),
   _target(),
   _dispatcher( 0 ),
   _cacheCount( 0 ),
   _callId    ( 0 ),
   _receivers()
{
#ifdef _WIN32
   WSADATA wsaData;
   if( ! utilCheckSysCall( 0 ==
      WSAStartup( MAKEWORD( 2, 2 ), &wsaData ),
      __FILE__, __LINE__, "WSAStartup" ))
   {
      throw std::runtime_error( "WSAStartup" );
   }
#endif
   CPPCHK(__FILE__,__LINE__,
      ((_out = socket( AF_INET, SOCK_DGRAM, 0 ))<0)
         ? UTIL_STATUS_STD_API_ERROR : UTIL_STATUS_NO_ERROR)
   memset( &_target, 0, sizeof( _target ));
   _target.sin_family      = AF_INET;
   _target.sin_port        = htons( addr._port );
   _target.sin_addr.s_addr = inet_addr( addr._inetAddress.c_str());
   int trueValue = 1;
   CPPCHK(__FILE__, __LINE__,
      setsockopt( _out, SOL_SOCKET, SO_REUSEADDR, (char*)&trueValue, sizeof( trueValue ))
         ? UTIL_STATUS_STD_API_ERROR : UTIL_STATUS_NO_ERROR )
   struct in_addr lIntrfc;
   memset( &lIntrfc, 0, sizeof( lIntrfc ));
   lIntrfc.s_addr = inet_addr( intrfc.c_str());
   CPPCHK(__FILE__, __LINE__,
      setsockopt( _out, IPPROTO_IP, IP_MULTICAST_IF, (char *)&lIntrfc, sizeof( lIntrfc ))
         ? UTIL_STATUS_STD_API_ERROR : UTIL_STATUS_NO_ERROR )
   printf( "sending to %s:%d via interface %s\n",
      addr._inetAddress.c_str(), addr._port, intrfc.c_str());
   _dispatcher            = new Dispatcher( *this );
   _caches[_cacheCount++] = new Cache( *this );
}

ParticipantImpl:: ~ ParticipantImpl() {
   delete _dispatcher;
   for( byte i = 0U; i < _cacheCount; ++i ) {
      delete _caches[i];
   }
   _cacheCount = 0;
   for( networkReceiversIter_t it = _receivers.begin(); it != _receivers.end(); ++it ) {
      delete *it;
   }
}

void ParticipantImpl::listen( const IRegistry & registry, const std::string & networkInterface ) {
   const socketAddresses_t & participants = registry.getParticipants();
   for( socketAddressesCstIter_t it = participants.begin(); it != participants.end(); ++it ) {
      const io::InetSocketAddress & addr     = *it;
      NetworkReceiver *             receiver =
         new NetworkReceiver( *this, addr, networkInterface );
      _receivers.push_back( receiver );
   }
}

void ParticipantImpl::registerLocalFactory( const ClassID & id, localFactory_t factory ) {
   os::Synchronized sync( _factoriesMutex );
   _localFactories[id] = factory;
}

void ParticipantImpl::registerRemoteFactory( const ClassID & id, ICRUD * factory ) {
   os::Synchronized sync( _factoriesMutex );
   _remoteFactories[id] = factory;
}

ICache & ParticipantImpl::getDefaultCache() {
   os::Synchronized sync( _cachesMutex );
   return *_caches[0];
}

ICache & ParticipantImpl::createCache( byte & id ) {
   os::Synchronized sync( _cachesMutex );
   id = _cacheCount++;
   _caches[id] = new Cache( *this );
   return *_caches[id];
}

ICache & ParticipantImpl::getCache( byte id ) {
   os::Synchronized sync( _cachesMutex );
   return *_caches[id];
}

IDispatcher & ParticipantImpl::getDispatcher() {
   return *_dispatcher;
}

void ParticipantImpl::pushCreateOrUpdateItem( Shareable * item ) {
   os::Synchronized sync( _outMutex );
   _payload.clear();
   item->serialize( _payload );
   _payload.flip();
   unsigned size = (unsigned)_payload.remaining();
   _header.clear();
   _header.put( SIGNATURE, 0, SIGNATURE_SIZE );
   _header.putByte( DATA_CREATE_OR_UPDATE );
   _header.putUInt( size );
   item->_id   .serialize( _header );
   item->_class.serialize( _header );
   _header.flip();
   _message.clear();
   _message.put( _header  );
   _message.put( _payload );
   _message.flip();
   _message.sendTo( _out, _target );
}

void ParticipantImpl::publishUpdated( shareables_t & updated ) {
   for( shareablesIter_t it = updated.begin(); it != updated.end(); ++it ) {
      pushCreateOrUpdateItem( *it );
   }
}

void ParticipantImpl::pushDeleteItem( Shareable * item ) {
   os::Synchronized sync( _outMutex );
   _header.put( SIGNATURE, 0, SIGNATURE_SIZE );
   _header.putByte( DATA_DELETE );
   item->_id.serialize( _header );
   _header.flip();
   _header.sendTo( _out, _target );
}

void ParticipantImpl::publishDeleted( shareables_t & deleted ) {
   for( shareablesIter_t it = deleted.begin(); it != deleted.end(); ++it ) {
      pushDeleteItem( *it );
   }
}

void ParticipantImpl::call(
   const std::string & intrfcName,
   const std::string & opName,
   const Arguments *   args,
   int                 callId )
{
   os::Synchronized sync( _outMutex );
   _message.clear();
   _message.put      ( SIGNATURE, 0, SIGNATURE_SIZE );
   _message.putByte  ( OPERATION );
   _message.putString( intrfcName );
   _message.putString( opName );
   _message.putInt   ( callId );
   _message.putByte  ( args ? (byte)args->getCount() : 0 );
   if( args ) {
      args->serialize( _message );
   }
   _message.flip();
   _message.sendTo( _out, _target );
}

int ParticipantImpl::call(
   const std::string & intrfcName,
   const std::string & opName,
   const Arguments *   args,
   ICallback &         callback )
{
   call( intrfcName, opName, args, _callId );
   _callbacks[_callId] = &callback;
   return _callId++;
}

Shareable * ParticipantImpl::newInstance( const ClassID & classId, io::ByteBuffer & frame ) {
   os::Synchronized sync( _factoriesMutex );
   localFactoriesIter_t it = _localFactories.find( classId );
   if( it == _localFactories.end()) {
      return 0;
   }
   localFactory_t factory = it->second;
   if( ! factory ) {
      return 0;
   }
   Shareable * item = factory();
   item->unserialize( frame );
   return item;
}

void ParticipantImpl::dataDelete( const dcrud::GUID & id ) {
   os::Synchronized sync( _cachesMutex );
   for( unsigned i = 0U; i < CACHE_COUNT; ++i ) {
      Cache * cache = _caches[i];
      if( cache ) {
         cache->deleteFromNetwork( id );
      }
      else {
         break;
      }
   }
}

void ParticipantImpl::dataUpdate( io::ByteBuffer & frame, size_t payloadSize ) {
   os::Synchronized sync( _cachesMutex );
   for( unsigned i = 0U; i < CACHE_COUNT; ++i ) {
      Cache * cache = _caches[i];
      if( cache ) {
         cache->updateFromNetwork( frame.copy( GUID_SIZE + CLASS_ID_SIZE + payloadSize ));
      }
      else {
         break;
      }
   }
}

void ParticipantImpl::callback(
   const std::string & intrfcName,
   const std::string & opName,
   const Arguments &   results,
   int                 callId  )
{
   callbacksIter_t it = _callbacks.find( callId );
   if( it == _callbacks.end()) {
      fprintf( stderr, "Unknown Callback received: %s.%s, id: %d\n",
         intrfcName.c_str(), opName.c_str(), -callId );
   }
   else {
      ICallback * callback = it->second;
      if( callback ) {
         callback->callback( intrfcName, opName, results );
      }
      else {
         fprintf( stderr, "Unknown Callback received: %s.%s, id: %d\n",
            intrfcName.c_str(), opName.c_str(), -callId );
      }
   }
}

bool ParticipantImpl::create( const ClassID & clsId, const Arguments & how ) {
   os::Synchronized sync( _factoriesMutex );
   remoteFactoriesIter_t it = _remoteFactories.find( clsId );
   if( it == _remoteFactories.end()) {
      return false;
   }
   ICRUD * factory = it->second;
   if( ! factory ) {
      return false;
   }
   factory->create( how );
   return true;
}

bool ParticipantImpl::update( const dcrud::GUID & id, const Arguments & how ) {
   os::Synchronized sync( _factoriesMutex );
   for( int i = 0; i < _cacheCount; ++i ) {
      Shareable * item = _caches[i]->read( id );
      if( item ) {
         remoteFactoriesIter_t it = _remoteFactories.find( item->_class );
         if( it == _remoteFactories.end()) {
            return false;
         }
         ICRUD * factory = it->second;
         if( ! factory ) {
            return false;
         }
         factory->update( *item, how );
         return true;
      }
   }
   return false;
}

bool ParticipantImpl::deleTe( const dcrud::GUID & id ) {
   os::Synchronized sync( _factoriesMutex );
   for( int i = 0; i < _cacheCount; ++i ) {
      Shareable * item = _caches[i]->read( id );
      if( item ) {
         remoteFactoriesIter_t it = _remoteFactories.find( item->_class );
         if( it == _remoteFactories.end()) {
            return false;
         }
         ICRUD * factory = it->second;
         if( ! factory ) {
            return false;
         }
         factory->deleTe( *item );
         return true;
      }
   }
   return false;
}
